Create OpenSuggestionCommand and OpenSuggestionCommandTest

johannagwan opened the pull request:
Closes #142 

So far, in this PR:
- [x]  Implemented the functionality of `execute` method as stated in #142 
- [x] Updated `OpenSuggestionCommand` to take in our tree. However, the functionality of the `updatePossiblePaths` method will only work after `BlockModelImpl` has been created. This is because the method needs to access the `root` of our `BlockTree` in the model.
- [x] Added `OpenSuggestionCommandTest`

kevinputera commented:
@johannagwan Hey Johanna, your branch is out of date from the master branch. Let's update it so that you can use the `AbsolutePath` and `RelativePath` classes!

kevinputera commented:
> @johannagwan Hey Johanna, your branch is out of date from the master branch. Let's update it so that you can use the `AbsolutePath` and `RelativePath` classes!

Oops my bad! Turns out you are already using the new paths 😅

kevinputera commented on src/main/java/com/notably/logic/suggestion/commands/OpenSuggestionCommand.java:
```diff
@@ -0,0 +1,138 @@
+            suggestions.add(suggestionItem);
+        }
+
+        return suggestions;
```
Take a look at `Stream#map` [here](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/stream/Stream.html#map(java.util.function.Function)) to make your code above more [declarative](https://en.wikipedia.org/wiki/Declarative_programming)!

For this case, the above code can be rewritten as:
```java
return possiblePaths.stream()
  .map(path -> {
    String displayText = path.toString();
    Runnable action = () -> {
      model.setInput(displayText);
    };
    return new SuggestionItemImpl(displayText, action);
  })
  .collect(Collectors.toList());
```

johannagwan replied:
Wowza sickerino! Thanks kevin! I updated the code.

kevinputera commented on src/main/java/com/notably/logic/suggestion/commands/OpenSuggestionCommand.java:
```diff
@@ -0,0 +1,137 @@
+
+        List<String> components = path.getComponents();
+        List<Node<String>> children = new ArrayList<>();
+        int index = 0;
```
Let's declare this just before the while-loop below since you're not using it in the for-loop below. That way you can logically group your code

kevinputera commented on src/main/java/com/notably/logic/suggestion/commands/OpenSuggestionCommand.java:
```diff
@@ -0,0 +1,137 @@
+
+        if (children.isEmpty()) { // if the note doesn't have any children
+            Path newPath = AbsolutePath.fromComponent(components);
+            possiblePaths.add(newPath);
```
Why is this needed?
I believe you can just do `possiblePaths.add(path)` directly?

johannagwan commented:
> 1. Would you want to deal with all operations inside `OpenSuggestionCommand` in terms of `AbsolutePath`s only? Perhaps you'd like to see how the `Block`-related API looks like first before deciding, but in case you'd want to work only in absolutes, I think it's better to use `AbsolutePath` in your class.

Alright, I have changed all my path to be `AbsolutePath`.

> 2. How do you currently envision the `Path` object passed into `OpenSuggestionCommand` be generated? Does `SuggestionEngine` parse the user input first, from `String` -> `Path`, before creating an `OpenSuggestionCommand` and with the `Path`? I believe it's best if we can draw parallels with how `Parser` and `Command` work here.

Okay, I will create an `OpenSuggestionCommandParser` class and subsequently, update my `SuggestionEngineImplementation` (in PR #128, but will only do it after I am done implementing the `OpenSuggestionCommand` and `SearchSuggestionCommand` and their respective parsers) accordingly.

> 3. When a user type `/note`, I believe the `/note` item should also be shown in the suggestion. Currently, you'll either show `/note` or not depending on whether it has children. I believe this is pretty inconsistent. What do you think?

I was initially following our Mock-up design. When the user types in `open Notably/`, the suggestion list does not contain `Notably/`. And I agree with your point about the consistency! I have updated my code as such. Thank you!

> 4. You're currently only providing the user with suggestions to the first level children of the current input. That is, if the user input is `/nus`, you'll only generate suggestions for `/nus/cs2103` and `/nus/cs2101`, but not `/nus/cs2103/notably`, for example. I believe we should generate the whole list?

Sorry for not doing it in time last time, but I have added the `getChildRecursive` method to do this.

Thanks @kevinputera for your detailed review! Wooohoooo

kevinputera commented on src/main/java/com/notably/logic/suggestion/commands/OpenSuggestionCommand.java:
```diff
@@ -0,0 +1,122 @@
+    public static final String COMMAND_WORD = "open";
+    public static final String RESPONSE_MESSAGE = "Open a note";
+    private AbsolutePath path;
+    private List<AbsolutePath> possiblePaths = new ArrayList<>();
```
Let's add a newline above this group to logically separate the static fields from the instance fields?

kevinputera commented on src/main/java/com/notably/logic/suggestion/commands/OpenSuggestionCommand.java:
```diff
@@ -0,0 +1,122 @@
+
+    public void setResponseTextToModel(Model model) {
+        model.setResponseText(RESPONSE_MESSAGE);
+    }
```
Nit: I'm actually not convinced that this abstraction is necessary!

kevinputera commented on src/main/java/com/notably/logic/suggestion/commands/OpenSuggestionCommand.java:
```diff
@@ -0,0 +1,122 @@
+    public static final String COMMAND_WORD = "open";
+    public static final String RESPONSE_MESSAGE = "Open a note";
+    private AbsolutePath path;
+    private List<AbsolutePath> possiblePaths = new ArrayList<>();
```
Do you actually need to keep `possiblePaths` as a field? From my understanding, you don't have to

johannagwan replied:
I kept it as a field since I will mutate `possiblePaths` in my `getChildRecursive` method. But aight I will change it!

kevinputera commented on src/main/java/com/notably/logic/suggestion/commands/OpenSuggestionCommand.java:
```diff
@@ -0,0 +1,122 @@
+                    index++;
+                }
+            }
+        }
```
Actually, I think this can be simplified as follows right?

```java
int index = 0;
while (index < components.size() && !queue.isEmpty()) {
  BlockTreeItem curr = queue.poll();
  String dir = components.get(index);
  if (curr.getTitle().equals(dir)) {
    children = curr.getBlockChildren();

    queue.clear();
    queue.addAll(children);

    index++;
  }
}
```

kevinputera replied:
Another thing: I believe it should be `curr.getTitle().getText().equals(dir)` instead!

johannagwan replied:
yOz you are correct!! thanks kevin!!

kevinputera commented on src/main/java/com/notably/logic/suggestion/commands/OpenSuggestionCommand.java:
```diff
@@ -0,0 +1,122 @@
+     * @param path The user's input path.
+     * @param model The app's model.
+     */
+    public void updatePossiblePaths(AbsolutePath path, Model model) {
```
I believe the logic here is working fine! However, try considering the below's step-by-step alternative (which I believe is much simpler). Let me know what you think!

1. First, you get the `BlockTreeItem` instance representing the given `path`. You can do this by calling `model.getBlockTree().get(path)`.
2. Now, you can treat this `BlockTreeItem` as the ancestor of all the paths that you would like to retrieve. You can do a generic BFS through the whole tree, treating the `BlockTreeItem` from before as the root, and return all the paths available!

kevinputera commented on src/main/java/com/notably/logic/suggestion/commands/OpenSuggestionCommand.java:
```diff
@@ -0,0 +1,122 @@
+        setResponseTextToModel(model);
+        updatePossiblePaths(path, model);
+        List<AbsolutePath> possiblePaths = getPossiblePaths();
+        List<SuggestionItem> suggestions = getSuggestions(possiblePaths, model);
```
Hmmm, your `updatePossiblePaths` method currently mutates the `possiblePaths` field, which I feel sort of unintuitive considering how `possiblePaths` can be transient here. I think you don't have to keep `possiblePaths` as a state over here.

From a high level perspective, if you can achieve the following flow, I think things will look much better!
```java
public void execute(Model model) {
  // Set response text
  setResponseTextToModel(model);

  // Set suggestions
  List<AbsolutePath> possiblePaths = getPossiblePaths(); // Here you generate the list, don't need to store in this object's field
  List<SuggestionItem> suggestions = getSuggestions(possiblePaths, model);
  setSuggestionsToModel(model, suggestions);
}
```

kevinputera commented on src/test/java/com/notably/logic/suggestion/block/BlockTreeItemStub.java:
```diff
@@ -0,0 +1,97 @@
+/**
+ * A Stub class for BlockTreeItem.
+ */
+public class BlockTreeItemStub implements BlockTreeItem {
```
I don't think you need this stub! Firzan has already implemented `BlockTreeItemImpl`, so you can just create those

kevinputera commented on src/test/java/com/notably/logic/suggestion/block/BlockTreeStub.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * A Stub class for BlockTree.
+ */
+public class BlockTreeStub implements BlockTree {
```
Similarly, I don't think you need to create stubs for `BlockTree`. Firzan has created `BlockTreeImpl`, which you can instantiate directly!

kevinputera commented on src/test/java/com/notably/logic/suggestion/commands/OpenSuggestionCommandTest.java:
```diff
@@ -0,0 +1,133 @@
+package com.notably.logic.suggestion.commands;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+//import static org.junit.jupiter.api.Assertions.assertTrue;
```
Let's remove this altogether if it's not needed

kevinputera commented on src/test/java/com/notably/logic/suggestion/commands/OpenSuggestionCommandTest.java:
```diff
@@ -0,0 +1,133 @@
+    private static AbsolutePath toCs2103Week3;
+    private static AbsolutePath toCs2103Week1Lecture;
+    private Model model;
+    private OpenSuggestionCommand openSuggestionCommand;
```
Similar to before, let's add a newline between the two groupings of static and instance fields

kevinputera commented on src/test/java/com/notably/logic/suggestion/commands/OpenSuggestionCommandTest.java:
```diff
@@ -0,0 +1,133 @@
+    public void tearDownModel() {
+        model = new ModelStub();
+    }
+
```
Actually, you don't really need these right? Considering you've already reinitialized both whenever the equivalent @BeforeEach methods are called

johannagwan commented:
Thanks for the detailed review, @kevinputera !

> In addition, I think you should test the `execute` method of your `OpenSuggestionCommand`, which is the main point of interaction with the outside world, rather than testing the `setResponseToModel`, `setSuggestionsToModel`, and `getSuggestions` methods. These three methods that you are testing can be considered implementation details, and are very prone to change. [Read](https://kentcdodds.com/blog/how-to-know-what-to-test) this if you're interested!

Alright, I'm going to change my test cases. Thank you!

kevinputera commented on src/main/java/com/notably/logic/suggestion/commands/OpenSuggestionCommand.java:
```diff
@@ -0,0 +1,117 @@
+        Objects.requireNonNull(suggestions);
+
+        model.setSuggestions(suggestions);
+    }
```
Actually, my earlier point applies here too 😅
I think it's clearer if you just pass `model.setSuggestions(suggestions)` directly in `execute`? It's a one-liner in the end of the day

kevinputera commented on src/main/java/com/notably/logic/suggestion/commands/OpenSuggestionCommand.java:
```diff
@@ -0,0 +1,117 @@
+            }
+        }
+        return possiblePaths;
+    }
```
Nit: If you're already mutating the supplied `possiblePaths`, why would you still return `possiblePaths`? It seems a bit inconsistent to me 😅

kevinputera commented on src/test/java/com/notably/logic/suggestion/commands/OpenSuggestionCommandTest.java:
```diff
@@ -0,0 +1,108 @@
+    public void execute_nullModel_throwsNullPointerException() {
+        OpenSuggestionCommand openSuggestionCommand = new OpenSuggestionCommand(toRoot);
+        assertThrows(NullPointerException.class, () -> openSuggestionCommand.execute(null));
+    }
```
You should add at least one more test case for the `execute` method! In fact, if you ask me, I'd say don't need to test the other methods. Simply testing your `execute` method should suffice.

Consider this:
- You can test your `getPossiblePaths` method or your `getSuggestions` method all day and make sure it works. Nevertheless, if you never test your `execute` method thoroughly, you can never confidently say that your code works, since the user of your module uses the `execute` method only!
- On the other hand, if you test the `execute` method only, you can always be confident that your `execute` method, plus all the methods it uses (well, perhaps not all) works!

kevinputera commented on src/test/java/com/notably/logic/suggestion/model/ModelStub.java:
```diff
@@ -0,0 +1,94 @@
+            blockTree.add(toCs2103Week1, lecture);
+        } catch (InvalidPathException e) {
+            throw new AssertionError(e);
+        }
```
If you already set up your blockTree here, why do you still need to do so in your `OpenSuggestionCommandTest.java` file?

kevinputera commented on src/test/java/com/notably/logic/suggestion/model/ModelStub.java:
```diff
@@ -0,0 +1,94 @@
+/**
+ * A Stub class for Model.
+ */
+public class ModelStub extends ModelStubBase {
```
Perhaps it's better if you move this stub to be a `private static class` of the `OpenSuggestionCommandTest` class. You're not using it any other place anyways. In addition, both of these classes are deeply coupled to one another, so perhaps separating them is not the best idea

kevinputera replied:
I also don't really like how this class is under the `com.notably.logic.suggestion.model` package. Seems rather unintuitive to have a `model` subpackage inside a `logic`
